keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,639 @@
+However, bucket management lacks some of the notable requirements of block and file provisioning, such as no dependency on the kubelet, no node topology constraints, etc. This allows for an architecture with lower overall complexity.
+This proposal does _not_ include a standardized protocol or abstraction of storage vendor APIs.  
+
+## Motivation 
+
+File and block are first class citizens within the Kubernetes ecosystem.  Object, though different on a fundamental level and lacking an open, committee controlled interface like POSIX, is a popular means of storing data, especially against very large data sources.   As such, we feel it is in the interest of the community to elevate buckets to a community supported feature.  In doing so, we can provide Kubernetes cluster users and administrators a normalized and familiar means of managing object storage.
+
+## Goals
++ Define a control plane API in order to standardize and formalize Kubernetes bucket provisioning
++ Be un-opinionated about the underlying object-store.
++ Present similar workflows for both _greenfield_  and _brownfield_ bucket provisioning.
++ Minimize privileges required to run a storage driver.
++ Minimize technical ramp-up with a design that is familiar to CSI storage driver authors and Kubernetes storage admins.
++ Use standard Kubernetes mechanisms to sync a pod with the readiness of the bucket it will consume. This can be accomplished via Secrets.
+
+## Non-Goals
++ Define a native _data-plane_ object store API which would greatly improve object store app portability.
++ Mirror the static workflow of PersistentVolumes wherein users are given the first available Volume.  Pre-provisioned buckets are expected to be non-empty and thus unique.
+
+##  Vocabulary
+
++  _Brownfield Bucket_ - externally created and represented by a `BucketClass` and managed by a provisioner.
++ _Bucket_ - A user-namespaced custom resource representing an object store bucket.
++  _BucketClass_ - A cluster-scoped custom resource containing fields defining the provisioner and an immutable parameter set.
+   + _In Greenfield_: an abstraction of new bucket provisioning.
+   + _In Brownfield_: an abstration of an existing objet store bucket.
++ _BucketContent_ - A cluster-scoped custom resource bound to a `Bucket` and containing relevant metadata.
++ _Container Object Storage Interface (COSI)_ -  A specification of gRPC data and methods making up the communication protocol between the driver and the sidecar.
++ _COSI Controller_ - A central controller responsible for managing `Buckets`, `BucketContents`, and Secrets.
++ _Driver_ - A containerized gRPC server which implements a storage vendor’s business logic through the COSI interface. It can be written in any language supported by gRPC and is independent of Kubernetes.
```

Comment:
Why is a grpc interface needed?  Just use a controller.

Grpc plug-ins lack many desirable characteristics of controllers and can be almost completely eliminated for the sorts of operations described here.

Reply:
Our motivation for using gRPC is similar to CSI's reasoning: 1) allow storage vendor provisioning to be container orchestrator agnostic. 2) allow storage vendors to write their provisioners in a variety of languages with the potential that the same driver could run in different orchestrators. 3) separate patching/versioning of vendor-specific code from k8s controller/sidecar code (which is difficult if a library is to be imported by the vendor).
The downsides are 1) there are separate containers now: a) the driver itself, b) a sidecar container mapping k8s/go structs to gRPC services/messages, c) a central controller. 2) it's more complex than a single controller.

What controller characteristics are we foregoing with this proposal? Thanks.

Reply:
I don't think CSI needed to use GRPC except at a node level, and honestly most of the node level things are using GRPC because we don't have good support for controller apis at the node level.  That's not a reason to use GRPC otherplaces.

Yeah, I would say 1 doesn't seem to be the right approach (generic things to container orchestrators is less critical to success of kube than it used to be).

2 you can write controllers in lots of languages, but supporting lots of languages is not really a technical design goal of good software in general, it is more specific than that (sometimes supporting lots of languages is valuable).  How many CSI drivers are in a non-go language?  How many CRI runtimes?  Either way, controllers can be in many languages and that is just as important a goal as anything else.

3. I don't buy the argument that vendoring kube code is hard - it was 4 years ago, it shouldn't be now.  Our ecosystem is much expanded.

This is the classic controller pattern - talk to a backend, provision something, update an api object.  You can collaborate with multiple controllers just as easily as with GRPC plugins.  You reduce the total stack complexity.  You can define a stock Go interface that pepole can fork.  You can describe the pattern for a third party.

I am very anti-GRPC style plugins when controllers solve the problem equally well.  We have allowed it in auth plugins and a few node level things, but it points to gaps in tooling we have mostly closed at this point.  Most plugins need to make kube api calls, so they aleady inherit most of a controller.

Reply:
Just reviewing the KEP, i see nothing that requires a gRPC plugin, I just see a bog standard kubernetes controller (which is our preferred mechanism for 3rd party extension).  Having an optional GRPC plugin for a small set of vendors sounds great, but since every major cloud and storage platform I'm aware of already has a go object storage bucket API (either as part of an operator or using a small integration), using go does not seem to be a blocker.

The dominant pattern should be a set of base code and Go interfaces for implementing those with the ability to both keep the core libraries up to date and fork those to provide additional behavior.

Reply:
Thanks @smarterclayton.  These are great points.  1 and 2 I agree with outright.

> I don't buy the argument that vendoring kube code is hard - it was 4 years ago, it shouldn't be now. Our ecosystem is much expanded.

IMO, this has been the biggest reason we've been amicable towards gRPC over a go interface.  We went down this path in a [previous incarnation](https://github.com/kube-object-storage/lib-bucket-provisioner) of a bucket provisioning API.  

The issue isn't that importing k8s tools is hard, it's when multiple dependencies wraps k8s tools, they inevitably result in a dep hell.  Go mods alleviate this for the most part but I don't think we can expect existing projects, particularly those written prior to mods, to _not_ encounter the same pain we did before.

> The dominant pattern should be a set of base code and Go interfaces for implementing those with the ability to both keep the core libraries up to date and fork those to provide additional behavior.

What I think you're suggesting is a that _instead_ of a library, a standalone controller be provided along with a go interface.  In effect, merging the sidecar and provisioner code under a single binary.  The controller manages the APIs and executes provisioner methods, deployed alone in it's one container.  Am I close to the mark?